File: shells/vscode/src/handoff.ts
import { createHash } from "node:crypto";

export const LOCAL_HANDOFF_NOTICE =
  "This packet remains local. AgentMage has not contacted Codex or any external service. External handling begins only if you manually transfer selected content.";

export type HandoffProhibitedAction =
  | "codex_invocation"
  | "tab_activation"
  | "chat_population"
  | "clipboard_write"
  | "uri_launch"
  | "local_runtime_delivery"
  | "raw_runtime_delivery"
  | "network_call"
  | "automatic_submission";

export interface HandoffPacketManifest {
  readonly schema_version: 2;
  readonly handoff_id: string;
  readonly draft_sha256: string;
  readonly entry_sha256: readonly string[];
  readonly packet_sha256: string;
  readonly packet_bytes: number;
  readonly destination: "manual_codex_interface";
  readonly acknowledgment_required: boolean;
  readonly delivered: false;
  readonly manifest_sha256: string;
}

export interface HandoffReview {
  readonly schema_version: 2;
  readonly preview_id: string;
  readonly packet_markdown: string;
  readonly manifest: HandoffPacketManifest;
  readonly local_only_notice: string;
  readonly expires_at_ms: number;
  readonly confirmation_sha256: string;
}

export interface LocalHandoffReceipt {
  readonly schema_version: 2;
  readonly attempt_id: string;
  readonly handoff_id: string | null;
  readonly outcome: "rendered" | "cancelled" | "denied";
  readonly result_code: string;
  readonly prohibited_action: HandoffProhibitedAction | null;
  readonly packet_sha256: string | null;
  readonly external_delivery_attempted: false;
  readonly receipt_sha256: string;
}

export interface RenderedHandoff {
  readonly schema_version: 2;
  readonly packet_markdown: string;
  readonly manifest: HandoffPacketManifest;
  readonly receipt: LocalHandoffReceipt;
}

const SHA256 = /^[0-9a-f]{64}$/;
const IDENTIFIER = /^[A-Za-z0-9][A-Za-z0-9._:-]{0,127}$/;
const CODE = /^[a-z0-9][a-z0-9._-]{0,127}$/;
const ACTIONS: readonly HandoffProhibitedAction[] = [
  "codex_invocation",
  "tab_activation",
  "chat_population",
  "clipboard_write",
  "uri_launch",
  "local_runtime_delivery",
  "raw_runtime_delivery",
  "network_call",
  "automatic_submission",
];

export function parseHandoffReview(candidate: unknown): HandoffReview {
  const value = record(candidate);
  requireKeys(value, [
    "confirmation_sha256",
    "expires_at_ms",
    "local_only_notice",
    "manifest",
    "packet_markdown",
    "preview_id",
    "schema_version",
  ]);
  const manifest = parseManifest(value.manifest);
  if (
    value.schema_version !== 2 ||
    !identifier(value.preview_id) ||
    typeof value.packet_markdown !== "string" ||
    Buffer.byteLength(value.packet_markdown, "utf8") === 0 ||
    Buffer.byteLength(value.packet_markdown, "utf8") > 128 * 1024 ||
    value.local_only_notice !== LOCAL_HANDOFF_NOTICE ||
    !positiveInteger(value.expires_at_ms) ||
    !sha(value.confirmation_sha256) ||
    manifest.packet_bytes !==
      Buffer.byteLength(value.packet_markdown, "utf8") ||
    manifest.packet_sha256 !== digestText(value.packet_markdown)
  ) {
    throw new Error("handoff.review.invalid");
  }
  const review: HandoffReview = {
    schema_version: 2,
    preview_id: value.preview_id,
    packet_markdown: value.packet_markdown,
    manifest,
    local_only_notice: LOCAL_HANDOFF_NOTICE,
    expires_at_ms: value.expires_at_ms,
    confirmation_sha256: value.confirmation_sha256,
  };
  const { confirmation_sha256: expected, ...unsigned } = review;
  if (digest(unsigned) !== expected) {
    throw new Error("handoff.review.invalid");
  }
  return review;
}

export function parseRenderedHandoff(candidate: unknown): RenderedHandoff {
  const value = record(candidate);
  requireKeys(value, [
    "manifest",
    "packet_markdown",
    "receipt",
    "schema_version",
  ]);
  const manifest = parseManifest(value.manifest);
  const receipt = parseLocalHandoffReceipt(value.receipt);
  if (
    value.schema_version !== 2 ||
    typeof value.packet_markdown !== "string" ||
    Buffer.byteLength(value.packet_markdown, "utf8") !==
      manifest.packet_bytes ||
    digestText(value.packet_markdown) !== manifest.packet_sha256 ||
    receipt.outcome !== "rendered" ||
    receipt.packet_sha256 !== manifest.packet_sha256 ||
    receipt.handoff_id !== manifest.handoff_id ||
    receipt.prohibited_action !== null
  ) {
    throw new Error("handoff.rendered.invalid");
  }
  return {
    schema_version: 2,
    packet_markdown: value.packet_markdown,
    manifest,
    receipt,
  };
}

export function parseLocalHandoffReceipt(
  candidate: unknown,
): LocalHandoffReceipt {
  const value = record(candidate);
  requireKeys(value, [
    "attempt_id",
    "external_delivery_attempted",
    "handoff_id",
    "outcome",
    "packet_sha256",
    "prohibited_action",
    "receipt_sha256",
    "result_code",
    "schema_version",
  ]);
  if (
    value.schema_version !== 2 ||
    !identifier(value.attempt_id) ||
    !(value.handoff_id === null || identifier(value.handoff_id)) ||
    !oneOf(value.outcome, ["rendered", "cancelled", "denied"] as const) ||
    !code(value.result_code) ||
    !(
      value.prohibited_action === null ||
      oneOf(value.prohibited_action, ACTIONS)
    ) ||
    !(value.packet_sha256 === null || sha(value.packet_sha256)) ||
    value.external_delivery_attempted !== false ||
    !sha(value.receipt_sha256)
  ) {
    throw new Error("handoff.receipt.invalid");
  }
  const receipt = value as unknown as LocalHandoffReceipt;
  const { receipt_sha256: expected, ...unsigned } = receipt;
  if (digest(unsigned) !== expected) {
    throw new Error("handoff.receipt.invalid");
  }
  return receipt;
}

function parseManifest(candidate: unknown): HandoffPacketManifest {
  const value = record(candidate);
  requireKeys(value, [
    "acknowledgment_required",
    "delivered",
    "destination",
    "draft_sha256",
    "entry_sha256",
    "handoff_id",
    "manifest_sha256",
    "packet_bytes",
    "packet_sha256",
    "schema_version",
  ]);
  if (
    value.schema_version !== 2 ||
    !identifier(value.handoff_id) ||
    !sha(value.draft_sha256) ||
    !Array.isArray(value.entry_sha256) ||
    value.entry_sha256.length === 0 ||
    value.entry_sha256.length > 32 ||
    !value.entry_sha256.every(sha) ||
    new Set(value.entry_sha256).size !== value.entry_sha256.length ||
    !sha(value.packet_sha256) ||
    !positiveInteger(value.packet_bytes) ||
    value.packet_bytes > 128 * 1024 ||
    value.destination !== "manual_codex_interface" ||
    typeof value.acknowledgment_required !== "boolean" ||
    value.delivered !== false ||
    !sha(value.manifest_sha256)
  ) {
    throw new Error("handoff.manifest.invalid");
  }
  const manifest = value as unknown as HandoffPacketManifest;
  const { manifest_sha256: expected, ...unsigned } = manifest;
  if (digest(unsigned) !== expected) {
    throw new Error("handoff.manifest.invalid");
  }
  return manifest;
}

function digest(value: unknown): string {
  return createHash("sha256")
    .update(JSON.stringify(value), "utf8")
    .digest("hex");
}

function digestText(value: string): string {
  return createHash("sha256").update(value, "utf8").digest("hex");
}

function record(value: unknown): Record<string, unknown> {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    throw new Error("handoff.shape.invalid");
  }
  return value as Record<string, unknown>;
}

function requireKeys(
  value: Record<string, unknown>,
  expected: readonly string[],
): void {
  const actual = Object.keys(value).sort();
  const ordered = [...expected].sort();
  if (
    actual.length !== ordered.length ||
    actual.some((item, index) => item !== ordered[index])
  ) {
    throw new Error("handoff.shape.invalid");
  }
}

function oneOf<const T extends string>(
  value: unknown,
  choices: readonly T[],
): value is T {
  return typeof value === "string" && choices.includes(value as T);
}

function identifier(value: unknown): value is string {
  return typeof value === "string" && IDENTIFIER.test(value);
}

function sha(value: unknown): value is string {
  return typeof value === "string" && SHA256.test(value);
}

function code(value: unknown): value is string {
  return typeof value === "string" && CODE.test(value);
}

function positiveInteger(value: unknown): value is number {
  return Number.isSafeInteger(value) && (value as number) > 0;
}

File: kernel/contracts/src/handoff.rs
//! Local-only, reviewable manual handoff contracts.

/// Sensitivity visible for one proposed handoff entry.
#[derive(Clone, Copy, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HandoffSensitivity {
    /// Public content that requires no additional acknowledgment.
    Public,
    /// Content supplied directly by the user.
    UserProvided,
    /// Permitted non-public content.
    NonPublic,
    /// Content that cannot enter a handoff packet.
    Prohibited,
}

/// Whether an entry is direct source material or an explicit inference.
#[derive(Clone, Copy, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HandoffEntryKind {
    /// Direct bounded source excerpt.
    SourceExcerpt,
    /// Explicitly labeled inferred content.
    Inference,
}

/// Redaction disposition applied before review.
#[derive(Clone, Copy, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HandoffEntryDisposition {
    /// The exact excerpt may be shown in the review and packet.
    Include,
    /// Only a fixed redaction marker may be shown.
    Redacted,
    /// The entry cannot enter a preview or packet.
    Prohibited,
}

/// Destination disclosed to the user without conferring delivery authority.
#[derive(Clone, Copy, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HandoffDestinationClass {
    /// Content may be manually moved by the user to a separate Codex interface.
    ManualCodexInterface,
}

/// One exact source, inference, or redaction disclosed before rendering.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub struct HandoffDisclosureEntry {
    /// Stable entry identity within the draft.
    pub entry_id: String,
    /// Stable source identity.
    pub source_id: String,
    /// Display-only relative file name or logical source label.
    pub display_source: String,
    /// Exact range, fragment, or logical excerpt identity.
    pub fragment: String,
    /// Exact reviewed excerpt or fixed redaction marker.
    pub excerpt: String,
    /// Digest of the complete observed source object.
    pub content_sha256: String,
    /// Direct-source or inference label.
    pub kind: HandoffEntryKind,
    /// Visible sensitivity label.
    pub sensitivity: HandoffSensitivity,
    /// Include, redact, or prohibit disposition.
    pub disposition: HandoffEntryDisposition,
    /// Ordered stable redaction codes.
    pub redactions: Vec<String>,
    /// Whether the source was hidden from ordinary workspace visibility.
    pub hidden: bool,
    /// Whether the entry is directly related to the declared objective.
    pub related: bool,
    /// Digest of all preceding entry fields.
    pub entry_sha256: String,
}

/// Complete local input proposed for a manual handoff.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub struct HandoffDraft {
    /// Contract schema version.
    pub schema_version: u16,
    /// Stable handoff identity.
    pub handoff_id: String,
    /// Current workspace-state digest.
    pub workspace_state_sha256: String,
    /// Current governing policy digest.
    pub policy_sha256: String,
    /// Current redaction-policy digest.
    pub redaction_policy_sha256: String,
    /// Exact objective.
    pub objective: String,
    /// Ordered acceptance criteria.
    pub acceptance_criteria: Vec<String>,
    /// Ordered authority, behavior, and output constraints.
    pub constraints: Vec<String>,
    /// Exact included source and inference disclosures.
    pub entries: Vec<HandoffDisclosureEntry>,
    /// Ordered, visible exclusions.
    pub exclusions: Vec<String>,
    /// Ordered unresolved questions.
    pub unresolved_questions: Vec<String>,
    /// Non-authoritative destination disclosure.
    pub destination: HandoffDestinationClass,
}

/// Content-addressed manifest for one exact rendered packet.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub struct HandoffPacketManifest {
    /// Contract schema version.
    pub schema_version: u16,
    /// Exact handoff identity.
    pub handoff_id: String,
    /// Digest of the source draft.
    pub draft_sha256: String,
    /// Ordered entry digests.
    pub entry_sha256: Vec<String>,
    /// Digest of exact UTF-8 packet bytes.
    pub packet_sha256: String,
    /// Exact UTF-8 packet size.
    pub packet_bytes: u64,
    /// Destination shown to the user.
    pub destination: HandoffDestinationClass,
    /// Whether permitted non-public content requires acknowledgment.
    pub acknowledgment_required: bool,
    /// Always false because AgentMage cannot deliver a handoff.
    pub delivered: bool,
    /// Digest of all preceding manifest fields.
    pub manifest_sha256: String,
}

/// Mandatory exact-content review shown before local rendering completes.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub struct HandoffReview {
    /// Contract schema version.
    pub schema_version: u16,
    /// Exact preview identity.
    pub preview_id: String,
    /// Exact packet text the user may later transfer manually.
    pub packet_markdown: String,
    /// Exact packet manifest.
    pub manifest: HandoffPacketManifest,
    /// Explicit local-only statement.
    pub local_only_notice: String,
    /// Logical expiry instant in milliseconds.
    pub expires_at_ms: u64,
    /// Digest bound to every review field and the required acknowledgment state.
    pub confirmation_sha256: String,
}

/// Prohibited delivery or interface-control action.
#[derive(Clone, Copy, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum HandoffProhibitedAction {
    /// Invoke Codex directly.
    CodexInvocation,
    /// Activate or control another tab.
    TabActivation,
    /// Populate another Chat interface.
    ChatPopulation,
    /// Write packet content to the clipboard.
    ClipboardWrite,
    /// Launch a URI or external handler.
    UriLaunch,
    /// Deliver through a local model endpoint.
    LocalRuntimeDelivery,
    /// Deliver through a raw runtime endpoint.
    RawRuntimeDelivery,
    /// Contact any network destination.
    NetworkCall,
    /// Submit the packet automatically.
    AutomaticSubmission,
    /// Upload packet content to any service or destination.
    FileUpload,
    /// Control a browser to deliver or populate packet content.
    BrowserControl,
    /// Deliver later from a schedule or timer.
    ScheduledDelivery,
    /// Route delivery through another model, tool, connector, or workflow.
    RoutedDelivery,
    /// Reuse standing consent instead of exact current user review.
    StandingConsentDelivery,
}

/// Local-only terminal outcome for a review or prohibited action attempt.
#[derive(Clone, Copy, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LocalHandoffOutcome {
    /// The current reviewed packet was rendered locally.
    Rendered,
    /// Rendering was cancelled.
    Cancelled,
    /// A stale, prohibited, or unacknowledged operation was denied.
    Denied,
}

/// Content-free local receipt that never claims external delivery.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub struct LocalHandoffReceipt {
    /// Contract schema version.
    pub schema_version: u16,
    /// Exact attempt identity.
    pub attempt_id: String,
    /// Exact handoff identity when a review existed.
    #[serde(deserialize_with = "crate::serialization::deserialize_required_option")]
    pub handoff_id: Option<String>,
    /// Terminal local outcome.
    pub outcome: LocalHandoffOutcome,
    /// Stable result or denial code.
    pub result_code: String,
    /// Prohibited action when one was attempted.
    #[serde(deserialize_with = "crate::serialization::deserialize_required_option")]
    pub prohibited_action: Option<HandoffProhibitedAction>,
    /// Exact packet digest when local rendering succeeded.
    #[serde(deserialize_with = "crate::serialization::deserialize_required_option")]
    pub packet_sha256: Option<String>,
    /// Always false; retained to prevent ambiguous delivery claims.
    pub external_delivery_attempted: bool,
    /// Digest of all preceding receipt fields.
    pub receipt_sha256: String,
}

/// Locally rendered packet, exact manifest, and no-delivery receipt.
#[derive(Clone, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
#[serde(deny_unknown_fields)]
pub struct RenderedHandoff {
    /// Contract schema version.
    pub schema_version: u16,
    /// Exact packet text previously reviewed.
    pub packet_markdown: String,
    /// Exact packet manifest previously reviewed.
    pub manifest: HandoffPacketManifest,
    /// Local-only terminal receipt.
    pub receipt: LocalHandoffReceipt,
}
